src/entry_point.rs

```diff
@@ -61,29 +85,83 @@ pub unsafe extern "C" fn _start(
         b .Lyield_loop
 
         .Lstack_init:
-        // Initialize the stack pointer. The stack pointer is computed as
+        // Compute the stacktop (stack_start). The stacktop is computed as
         // stack_size + mem_start plus padding to align the stack to a multiple
         // of 8 bytes. The 8 byte alignment is to follow ARM AAPCS:
         // http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.faqs/ka4127.html
-        ldr ip, [r0, #36]  // ip = text_start->stack_size
-        add ip, ip, r1     // ip = text_start->stack_size + mem_start
-        add ip, #7         // ip = text_start->stack_size + mem_start + 7
-        bic r1, ip, #7     // r1 = (text_start->stack_size + mem_start + 7) & ~0x7
-        mov sp, r1         // sp = r1
-
-        // Call rust_start. text_start, stack_top, and app_heap_break are
-        // already in the correct registers.
+        ldr r4, [r0, #36]  // r4 = app_start->stack_size
+        add r4, r4, r1     // r4 = app_start->stack_size + mem_start
+        add r4, #7         // r4 = app_start->stack_size + mem_start + 7
+        bic r4, r4, #7     // r4 = (app_start->stack_size + mem_start + 7) & ~0x7
+        mov sp, r4         // sp = r4
+
+        // We need to pass app_start, stacktop and app_heap_break to rust_start.
+        // Temporarily store them in r6, r7 and r8
+        mov r6, r0
+        mov r7, sp
+
+        // Debug support, tell the kernel the stack location
+        //
+        // memop(10, stacktop)
+        // r7 contains stacktop
+        mov r0, #10
+        mov r1, r7
+        svc 4
+
+        // Debug support, tell the kernel the heap_start location
+        mov r0, r6
+        ldr r4, [r0, #24] // r4 = app_start->bss_start
+        ldr r5, [r0, #28] // r5 = app_start->bss_size
+        add r4, r4, r5    // r4 = bss_start + bss_size
+        //
+        // memop(11, r4)
+        mov r0, #11
+        mov r1, r4
+        svc 4
+
+        // Store heap_start (and soon to be app_heap_break) in r8
+        mov r8, r4
```

Comment:
Line 123 ("mov r8, r4"):

I think this is unnecessary. As far as I can tell, r4 is not touched again in this assembly block, so you can remove this and use r4 instead of r8 on lines 138 and 151.

Reply:
That's correct. Its just my style for writing assembly code. I don't try to optimize for register reuse. Instead by separating out state into high registers and on stack (if required) we can write blocks of assembly code that can easily refactored around if required.